File: MIDTERM-LAB-3/JavaScript/mp15_export_first_50.js
const fs = require("node:fs");
const path = require("node:path");
const readline = require("node:readline");

// Asks a question in the terminal and returns trimmed user input.
function askQuestion(rl, promptText) {
    return new Promise((resolve) => {
        rl.question(promptText, (answer) => resolve(answer.trim()));
    });
}

// Creates an input reader that supports both interactive and piped execution.
function createInputReader() {
    if (!process.stdin.isTTY) {
        const pipedData = fs.readFileSync(0, "utf8").split(/\r?\n/);
        let cursor = 0;

        return {
            ask(promptText) {
                process.stdout.write(promptText);
                const value = (pipedData[cursor] || "").trim();
                cursor += 1;
                if (value) {
                    process.stdout.write(`${value}\n`);
                }
                return Promise.resolve(value);
            },
            close() {
                // No readline interface to close in piped mode.
            },
        };
    }

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });

    return {
        ask(promptText) {
            return askQuestion(rl, promptText);
        },
        close() {
            rl.close();
        },
    };
}

// Parses one CSV line and keeps commas inside quoted values intact.
function parseCsvLine(line) {
    const values = [];
    let current = "";
    let inQuotes = false;

    for (let i = 0; i < line.length; i += 1) {
        const ch = line[i];

        if (ch === '"') {
            if (inQuotes && i + 1 < line.length && line[i + 1] === '"') {
                current += '"';
                i += 1;
            } else {
                inQuotes = !inQuotes;
            }
        } else if (ch === "," && !inQuotes) {
            values.push(current);
            current = "";
        } else {
            current += ch;
        }
    }

    values.push(current);
    return values;
}

// Checks if all fields in the row are empty.
function isEmptyRow(row) {
    return row.every((value) => value.trim() === "");
}

// Pads or trims rows to match the header length.
function normalizeRowSize(row, targetSize) {
    const normalized = [...row];
    while (normalized.length < targetSize) {
        normalized.push("");
    }
    return normalized.slice(0, targetSize);
}

// Reads CSV, finds header, and returns parsed rows.
function loadDataset(datasetPath) {
    const raw = fs.readFileSync(datasetPath, "utf8");
    const lines = raw.split(/\r?\n/);

    let header = null;
    const rows = [];

    for (const line of lines) {
        const parsed = parseCsvLine(line);

        if (!header) {
            if (parsed.length > 0 && parsed[0].trim().toLowerCase() === "candidate") {
                header = [...parsed];
            }
            continue;
        }

        if (isEmptyRow(parsed)) {
            continue;
        }

        rows.push(normalizeRowSize(parsed, header.length));
    }

    if (!header) {
        throw new Error("Header row not found. Expected first column to be 'Candidate'.");
    }

    return { header, rows };
}

// Escapes one value for valid CSV output.
function escapeCsv(value) {
    const needsQuotes = value.includes(",") || value.includes('"') || value.includes("\n") || value.includes("\r");
    if (!needsQuotes) {
        return value;
    }
    return `"${value.replace(/"/g, '""')}"`;
}

// Writes CSV content to file using header + first N rows.
function exportRows(dataset, outputPath, rowCount) {
    const outputLines = [];
    outputLines.push(dataset.header.map(escapeCsv).join(","));

    for (let i = 0; i < rowCount; i += 1) {
        outputLines.push(dataset.rows[i].map(escapeCsv).join(","));
    }

    fs.writeFileSync(outputPath, `${outputLines.join("\n")}\n`, "utf8");
}

// Compares paths in normalized absolute form to avoid accidental overwrite.
function isSamePath(pathA, pathB) {
    const normalize = (p) => path.resolve(p).replace(/\//g, "\\").toLowerCase();
    return normalize(pathA) === normalize(pathB);
}

async function main() {
    const inputReader = createInputReader();

    try {
        // Required by instruction: ask dataset path first.
        const datasetPath = await inputReader.ask("Enter CSV dataset file path: ");
        if (!datasetPath) {
            console.log("Dataset path is required.");
            return;
        }

        // Auto-create output path inside MIDTERM-LAB-3/JavaScript as requested.
        const outputPath = path.join(__dirname, "mp15_first50_js_output.csv");

        if (isSamePath(datasetPath, outputPath)) {
            console.log("Output file must be different from dataset file.");
            console.log("Please provide a new CSV file path for export.");
            return;
        }

        const dataset = loadDataset(datasetPath);
        const exportCount = Math.min(50, dataset.rows.length);
        exportRows(dataset, outputPath, exportCount);

        // Formatted result summary.
        console.log();
        console.log("MP15 RESULT");
        console.log("-----------");
        console.log(`Total valid data rows: ${dataset.rows.length}`);
        console.log(`Rows exported       : ${exportCount}`);
        console.log(`Output file (auto)  : ${outputPath}`);
        console.log("Export completed successfully.");
    } catch (error) {
        console.log(`File processing error: ${error.message}`);
    } finally {
        inputReader.close();
    }
}

main();

File: MIDTERM-LAB-3/Java/MP15ExportFirst50.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

/**
 * MP15 - Export first 50 rows to CSV.
 * This program asks for a dataset path first, reads and parses the CSV,
 * then exports the first 50 valid rows (plus header) to a new CSV file.
 */
public class MP15ExportFirst50 {

    // Simple container for header and data rows after parsing.
    private static class Dataset {
        List<String> header;
        List<List<String>> rows;

        Dataset(List<String> header, List<List<String>> rows) {
            this.header = header;
            this.rows = rows;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        try {
            // Required by instruction: ask for dataset path before any processing.
            System.out.print("Enter CSV dataset file path: ");
            String datasetPath = scanner.nextLine().trim();
            if (datasetPath.isEmpty()) {
                System.out.println("Dataset path is required.");
                return;
            }

            File datasetFile = new File(datasetPath).getCanonicalFile();
            // Auto-create output path inside MIDTERM-LAB-3/Java as requested.
            File datasetParent = datasetFile.getParentFile();
            File outputFile;
            if (datasetParent != null) {
                outputFile = new File(new File(datasetParent, "Java"), "mp15_first50_java_output.csv");
            } else {
                outputFile = new File("mp15_first50_java_output.csv");
            }
            String outputPath = outputFile.getCanonicalPath();

            // Guard against overwriting the original dataset file.
            if (datasetFile.equals(new File(outputPath).getCanonicalFile())) {
                System.out.println("Output file must be different from dataset file.");
                System.out.println("Please provide a new CSV file path for export.");
                return;
            }

            Dataset dataset = loadDataset(datasetPath);
            int exportCount = Math.min(50, dataset.rows.size());
            exportRows(dataset, outputPath, exportCount);

            // Formatted summary output for easier checking.
            System.out.println();
            System.out.println("MP15 RESULT");
            System.out.println("-----------");
            System.out.println("Total valid data rows: " + dataset.rows.size());
            System.out.println("Rows exported       : " + exportCount);
            System.out.println("Output file (auto)  : " + outputPath);
            System.out.println("Export completed successfully.");
        } catch (IOException ex) {
            System.out.println("File processing error: " + ex.getMessage());
        } catch (Exception ex) {
            System.out.println("Unexpected error: " + ex.getMessage());
        } finally {
            scanner.close();
        }
    }

    // Loads CSV, skips metadata lines, finds the header row, and stores valid records.
    private static Dataset loadDataset(String datasetPath) throws IOException {
        List<String> header = new ArrayList<>();
        List<List<String>> rows = new ArrayList<>();
        boolean headerFound = false;

        try (BufferedReader reader = new BufferedReader(new FileReader(datasetPath))) {
            String line;
            while ((line = reader.readLine()) != null) {
                List<String> parsed = parseCsvLine(line);

                if (!headerFound) {
                    if (!parsed.isEmpty() && "Candidate".equalsIgnoreCase(parsed.get(0).trim())) {
                        header = new ArrayList<>(parsed);
                        headerFound = true;
                    }
                    continue;
                }

                if (isEmptyRow(parsed)) {
                    continue;
                }

                rows.add(normalizeRowSize(parsed, header.size()));
            }
        }

        if (!headerFound) {
            throw new IOException("Header row not found. Expected first column to be 'Candidate'.");
        }

        return new Dataset(header, rows);
    }

    // Exports header + first N rows into a CSV file.
    private static void exportRows(Dataset dataset, String outputPath, int rowCount) throws IOException {
        File outputFile = new File(outputPath);
        File parent = outputFile.getParentFile();
        if (parent != null && !parent.exists()) {
            parent.mkdirs();
        }

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(outputFile))) {
            writeCsvRow(writer, dataset.header);
            for (int i = 0; i < rowCount; i++) {
                writeCsvRow(writer, dataset.rows.get(i));
            }
        }
    }

    // Parses one CSV line with quote-aware logic to preserve commas inside quoted values.
    private static List<String> parseCsvLine(String line) {
        List<String> result = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        boolean inQuotes = false;

        for (int i = 0; i < line.length(); i++) {
            char ch = line.charAt(i);

            if (ch == '"') {
                if (inQuotes && i + 1 < line.length() && line.charAt(i + 1) == '"') {
                    current.append('"');
                    i++;
                } else {
                    inQuotes = !inQuotes;
                }
            } else if (ch == ',' && !inQuotes) {
                result.add(current.toString());
                current.setLength(0);
            } else {
                current.append(ch);
            }
        }

        result.add(current.toString());
        return result;
    }

    // Returns true if all fields are blank.
    private static boolean isEmptyRow(List<String> row) {
        for (String value : row) {
            if (!value.trim().isEmpty()) {
                return false;
            }
        }
        return true;
    }

    // Pads or trims rows to the same number of columns as the header.
    private static List<String> normalizeRowSize(List<String> row, int targetSize) {
        List<String> normalized = new ArrayList<>(row);
        while (normalized.size() < targetSize) {
            normalized.add("");
        }
        if (normalized.size() > targetSize) {
            normalized = new ArrayList<>(normalized.subList(0, targetSize));
        }
        return normalized;
    }

    // Writes one row to CSV with proper escaping.
    private static void writeCsvRow(BufferedWriter writer, List<String> row) throws IOException {
        for (int i = 0; i < row.size(); i++) {
            if (i > 0) {
                writer.write(',');
            }
            writer.write(escapeCsv(row.get(i)));
        }
        writer.newLine();
    }

    // Escapes a CSV value when it contains comma, quote, or line break.
    private static String escapeCsv(String value) {
        boolean needsQuotes = value.contains(",") || value.contains("\"") || value.contains("\n") || value.contains("\r");
        if (!needsQuotes) {
            return value;
        }
        return '"' + value.replace("\"", "\"\"") + '"';
    }
}
